src/js/types/masonry.js

```diff
@@ -11,49 +11,56 @@ export default class Masonry extends BaseGalleryType {
 		this.currentBreakpoint = currentBreakpoint;
 
 		const heights = [],
+			itemsInColumn = [],
 			aggregatedHeights = [],
 			columns = this.getCurrentDeviceSetting( 'columns' ),
 			containerWidth = this.$container.width(),
 			horizontalGap = this.getCurrentDeviceSetting( 'horizontalGap' ),
 			itemWidth = ( containerWidth - ( horizontalGap * ( columns - 1 ) ) ) / columns,
 			$items = this.getActiveItems();
+		let columnHeight = 0;
+
+		for ( let i = 0; i < columns; i++ ) {
+			itemsInColumn[ i ] = 0;
+			heights[ i ] = 0;
+		}
 
 		$items.each( ( index, item ) => {
-			const row = Math.floor( index / columns ),
-				indexAtRow = index % columns,
-				imageData = this.getImageData( index ),
+			const imageData = this.getImageData( index ),
 				itemHeight = itemWidth / imageData.ratio;
+			let indexAtRow = index % columns;
 
-			item.style.setProperty( '--item-height', ( imageData.height / imageData.width * 100 ) + '%' );
-			item.style.setProperty( '--column', indexAtRow );
+			columnHeight = heights[ indexAtRow ];
+			jQuery.each( heights, ( colNumber, colHeight ) => {
+				if ( ( colHeight > 5 ) && columnHeight > colHeight + 5 ) {
+					columnHeight = colHeight;
+					indexAtRow = colNumber;
+				}
+			} );
 
-			if ( row ) {
-				aggregatedHeights[ index ] = heights[ indexAtRow ];
+			aggregatedHeights[ index ] = heights[ indexAtRow ];
+			heights[ indexAtRow ] += itemHeight;
 
-				heights[ indexAtRow ] += itemHeight;
-			} else {
-				heights.push( itemHeight );
-			}
+			item.style.setProperty( '--item-height', ( imageData.height / imageData.width * 100 ) + '%' );
+			item.style.setProperty( '--column', indexAtRow );
+			item.style.setProperty( '--items-in-column', itemsInColumn[ indexAtRow ] );
+			itemsInColumn[ indexAtRow ]++;
 		} );
 
 		const highestColumn = Math.max( ...heights ),
 			highestColumnIndex = heights.indexOf( highestColumn ),
-			rows = Math.floor( this.settings.items.length / columns ),
-			rowsRemainder = this.settings.items.length % columns,
-			highestColumnsGapsCount = rowsRemainder > highestColumnIndex ? rows : rows - 1,
+			rows = itemsInColumn[ highestColumnIndex ],
+			highestColumnsGapsCount = rows - 1,
 			containerAspectRatio = highestColumn / containerWidth;
 
 		this.$container[ 0 ].style.setProperty( '--columns', columns );
-		this.$container[ 0 ].style.setProperty( '--highest-column-gap-count', highestColumnsGapsCount );
+		this.$container[ 0 ].style.setProperty( '--highest-column-gap-count', highestColumnsGapsCount.toString() );
```

Comment:
Why did you need to convert it to a string?

Reply:
PhpStorm claims this function expects a string... 
Should I ignore it?

Reply:
Yes, you should ignore it. It's unnecessary.